File: src/main/java/com/fg/util/babylon/export/SnapshotAdapter.java
package com.fg.util.babylon.export;

import com.fg.util.babylon.entity.MessageFileContent;
import com.fg.util.babylon.entity.PropertiesMap;
import com.fg.util.babylon.entity.Snapshot;

import java.util.NoSuchElementException;

/**
 * Implements snapshot contracts by delegating to [Snapshot] without exposing its details.
 */
public class SnapshotAdapter implements TranslationSnapshotReadContract, TranslationSnapshotWriteContract {

    private final Snapshot snapshot;

    public SnapshotAdapter(Snapshot snapshot) {
        this.snapshot = snapshot;
    }

    @Override
    public boolean includesMsgFile(String msgFile) {
        return snapshot.getPropFileByFileName(msgFile) != null;
    }

    @Override
    public Integer registerMsgFile(String msgFilePath) {
        MessageFileContent messageFileContent = snapshot.getOrPutNewPropFileByFileName(msgFilePath);
        return messageFileContent.id;
    }

    @Override
    public boolean containsMessage(String msgKey, String msgFile) {
        if (!includesMsgFile(msgFile)) return false;
        return containsMessageInProps(msgKey, getTranslationProperties(msgFile));
    }

    private boolean containsMessageInProps(String msgKey, PropertiesMap msgProps) {
        // PropertiesMap replaces null values with empty strings internally
        return msgProps.containsKey(msgKey) && msgProps.get(msgKey) != null && !msgProps.get(msgKey).isEmpty();
    }

    private PropertiesMap getTranslationProperties(String msgFile) {
        if (!includesMsgFile(msgFile)) {
            // caller should have asked before
            throw new NoSuchElementException("Translation snapshot doesn't contain message file '$msgFile'.");
        }
        return snapshot.getPropFileByFileName(msgFile).properties;
    }

    @Override
    public String getLastMessageValue(String msgKey, String msgFile) {
        PropertiesMap msgProps = getTranslationProperties(msgFile);
        if (!containsMessageInProps(msgKey, msgProps)) {
            // caller should have asked before
            String errMsg = "No message for key '" + msgKey + "' in '" + msgFile +"' message file snapshot.";
            throw new NoSuchElementException(errMsg);
        }
        return msgProps.get(msgKey);
    }

}

File: src/main/java/com/fg/util/babylon/export/SnapshotAdapter.kt
package com.fg.util.babylon.export

import com.fg.util.babylon.entity.PropertiesMap
import com.fg.util.babylon.entity.Snapshot
import java.util.NoSuchElementException


/**
 * Delegates to [Snapshot] without exposing its details.
 */
class SnapshotAdapter(private val snapshot: Snapshot) : TranslationSnapshotReadContract, TranslationSnapshotWriteContract {

    override fun includesMsgFile(msgFile: String): Boolean =
            snapshot.getPropFileByFileName(msgFile) != null

    override fun registerMsgFile(msgFilePath: MsgFilePath): Int {
        val messageFileContent = snapshot.getOrPutNewPropFileByFileName(msgFilePath)!!
        return messageFileContent.id!!
    }

    override fun containsMessage(msgKey: MessageKey, msgFile: MsgFilePath): Boolean =
            containsMessageInProps(msgKey, getTranslationProperties(msgFile))

    private fun containsMessageInProps(msgKey: MessageKey, msgProps: PropertiesMap): Boolean =
            // PropertiesMap replaces null values with empty strings internally
            if (!msgProps.containsKey(msgKey)) false else !msgProps[msgKey].isNullOrEmpty()

    override fun getLastMessageValue(msgKey: MessageKey, msgFile: MsgFilePath): Message {
        val msgProps = getTranslationProperties(msgFile)
        if (!containsMessageInProps(msgKey, msgProps)) {
            // caller should have asked before
            throw NoSuchElementException("No message for key '$msgKey' in '$msgFile' message file snapshot.")
        }
        return msgProps[msgKey]
    }

    private fun getTranslationProperties(msgFile: MsgFilePath): PropertiesMap {
        if (!includesMsgFile(msgFile)) {
            // caller should have asked before
            throw NoSuchElementException("Translation snapshot doesn't contain message file '$msgFile'.")
        }
        return snapshot.getPropFileByFileName(msgFile)?.properties!!
    }

}